fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePartitioner.java

```diff
@@ -800,57 +833,121 @@ public CompletableFuture<RecordCursorContinuation> rebalancePartitions(RecordCur
         }
         return getAllPartitionMetaInfo(groupingKey).thenCompose(partitionInfos -> {
             // need to track the next partition id to use when creating a new one
-            int maxPartitionId = partitionInfos.stream().map(LucenePartitionInfoProto.LucenePartitionInfo::getId).max(Integer::compare).orElse(0);
+            RepartitioningContext repartitioningContext = new RepartitioningContext();
+            repartitioningContext.maxPartitionId = partitionInfos.stream().map(LucenePartitionInfoProto.LucenePartitionInfo::getId).max(Integer::compare).orElse(0);
             if (LOGGER.isTraceEnabled()) {
                 LOGGER.trace(partitionInfos.stream()
                         .sorted(Comparator.comparing(pi -> Tuple.fromBytes(pi.getFrom().toByteArray())))
                         .map(pi -> "pi[" + pi.getId() + "]@" + pi.getCount() + Tuple.fromBytes(pi.getFrom().toByteArray()) + "->" + Tuple.fromBytes(pi.getTo().toByteArray()))
                         .collect(Collectors.joining(", ", "Rebalancing partitions (group=" + groupingKey + "): ", "")));
             }
 
-            for (LucenePartitionInfoProto.LucenePartitionInfo partitionInfo : partitionInfos) {
-                if (partitionInfo.getCount() > indexPartitionHighWatermark) {
-                    // process one partition
-                    if (LOGGER.isDebugEnabled()) {
-                        LOGGER.debug(repartitionLogMessage("Repartitioning records", groupingKey, repartitionDocumentCount, partitionInfo));
+            for (int i = 0; i < partitionInfos.size(); i++) {
+                LucenePartitionInfoProto.LucenePartitionInfo partitionInfo = partitionInfos.get(i);
+
+                if (LOGGER.isDebugEnabled()) {
+                    LOGGER.debug(repartitionLogMessage("Repartitioning records", groupingKey, repartitionDocumentCount, partitionInfo));
+                }
+
+                final int currentPartitionCount = partitionInfo.getCount();
+
+                if (currentPartitionCount >= indexPartitionLowWatermark && currentPartitionCount <= indexPartitionHighWatermark) {
+                    // repartitioning not needed
+                    continue;
+                }
+
+                if (currentPartitionCount < indexPartitionLowWatermark) {
+                    // partitionInfos are ordered next -> previous
+                    LucenePartitionInfoProto.LucenePartitionInfo next = i == 0 ? null : partitionInfos.get(i - 1);
+                    LucenePartitionInfoProto.LucenePartitionInfo previous = i == partitionInfos.size() - 1 ? null : partitionInfos.get(i + 1);
+                    repartitioningContext.countToMove = Math.min(currentPartitionCount, repartitionDocumentCount);
+                    if (currentPartitionCount > repartitionDocumentCount) {
+                        repartitioningContext.countToMove++;
+                    } else {
+                        repartitioningContext.emptyingPartition = true;
                     }
```

Comment:
It _might_ be clearer to remove the `Math.min` and just set `countToMove` in both blocks
```suggestion
                    if (currentPartitionCount > repartitionDocumentCount) {
                        repartitioningContext.countToMove = repartitionDocumentCount + 1;
                    } else {
                        repartitioningContext.countToMove = currentPartitionCount;
                        repartitioningContext.emptyingPartition = true;
                    }
```